analytics/src/test/java/com/segment/analytics/EdgeFunctionMiddlewareTest.kt

```diff
@@ -0,0 +1,141 @@
+                                        return integrationBar
+                                    }
+
+                                    override fun key(): String {
+                                        return "bar"
+                                    }
+                                })
+                        .executor(MoreExecutors.newDirectExecutorService())
+    }
+
+    /** Edge Function Middleware Tests **/
+    @Test
+    @Throws(Exception::class)
+    fun edgeFunctionMiddlewareCanExist() {
+
+        val analytics = builder
+                .useEdgeFunctionMiddleware(Mockito.mock(JSMiddleware::class.java))
+                .build()
+
+        analytics.track("foo")
+
+        val privateEdgeFunctions: Field = Analytics::class.java.getDeclaredField("edgeFunctionMiddleware")
+        Assertions.assertThat(privateEdgeFunctions).isNotNull
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun edgeFunctionMiddlewareOverwrites() {
+
+        try {
+            val analytics = builder
+                    .useEdgeFunctionMiddleware(Mockito.mock(JSMiddleware::class.java))
+                    .useSourceMiddleware(
+                            Middleware { throw AssertionError("should not be invoked") })
+                    .build()
+
+            Assertions.fail("Should not reach this state")
+        } catch (exception: java.lang.Exception) {
+            Assertions.assertThat(exception).isInstanceOf(IllegalStateException::class.java)
+        }
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun edgeFunctionDestinationMiddlewareOverwrites() {
+
+        try {
+            val analytics = builder
+                    .useEdgeFunctionMiddleware(Mockito.mock(JSMiddleware::class.java))
+                    .useDestinationMiddleware("test",
+                            Middleware { throw AssertionError("should not be invoked") })
+                    .build()
+
+            Assertions.fail("Should not reach this state")
+        } catch (exception: java.lang.Exception) {
+            Assertions.assertThat(exception).isInstanceOf(IllegalStateException::class.java)
+        }
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun edgeFunctionBothMiddlewareOverwrites() {
+
+        try {
+            val analytics = builder
+                    .useEdgeFunctionMiddleware(Mockito.mock(JSMiddleware::class.java))
+                    .useDestinationMiddleware("test",
+                            Middleware { throw AssertionError("should not be invoked") })
+                    .useSourceMiddleware(
+                            Middleware { throw AssertionError("should not be invoked") })
+                    .build()
+
+            Assertions.fail("Should not reach this state")
+        } catch (exception: java.lang.Exception) {
+            Assertions.assertThat(exception).isInstanceOf(IllegalStateException::class.java)
+        }
+    }
```

Comment:
can we also add a test for the scenario where we add a source/destination middleware before edge functions and an exception is raised?